### ISSUE TITLE
Process 3 terminated with signal SIGSEGV

### ISSUE THREAD

**Issue by ajesujoba** [pytorch/fairseq#1720]
I got the error `process 3 terminated with signal SIGSEGV` while doing a multi-GPU training. What can I do to resolve this?

![sigseg](https://user-images.githubusercontent.com/12751379/74836669-1daa9980-5320-11ea-9bf6-51def1f53fc6.PNG)

`python3 train.py data-bin/iwslt14.tokenized.de-en --arch transformer --share-decoder-input-output-embed  --opt
imizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0     --lr 5e-4 --lr-scheduler inverse_sqrt --warmup-updates 4000     --dropout 0.3 --weight-decay 0.0001 --criteri
on label_smoothed_cross_entropy --label-smoothing 0.1   --max-sentences 10   -s de -t en  --dataset-impl raw --share-all-embeddings --share-decoder-input-output-embed`

**Comment by gvskalyan** [comment]
If this `python -c 'from fairseq import libnat; print(libnat.suggested_ed2_path([[1, 2, 3, 4]], [[1, 3, 4, 5, 6]], 0))'` gives an error, the following might be a solution.

Please refer [this](https://github.com/pytorch/fairseq/issues/1346) issue.
Follow [these](https://github.com/pytorch/fairseq/issues/1346#issuecomment-567516336) steps.

**Comment by ajesujoba** [comment]
The response I got when I tried the command you gave

![libnat](https://user-images.githubusercontent.com/12751379/74841229-a3cade00-5328-11ea-938f-7b98a783954e.PNG)

**Comment by gvskalyan** [comment]
As per this https://github.com/pytorch/pytorch/issues/32291, do pip install numpy and then test.

**Comment by ajesujoba** [comment]
It appears I was able to run `python -c 'from fairseq import libnat; print(libnat.suggested_ed2_path([[1, 2, 3, 4]], [[1, 3, 4, 5, 6]], 0))'` without having an error. 
![sigseg3](https://user-images.githubusercontent.com/12751379/74876791-9cc1c100-5364-11ea-81de-b4efc7f70cdc.PNG)
Then
![sigseg4](https://user-images.githubusercontent.com/12751379/74876801-a0edde80-5364-11ea-9662-edadc7f0bff5.PNG)
The error persist
![sigseg5](https://user-images.githubusercontent.com/12751379/74876852-bfec7080-5364-11ea-8b87-700d2b8217a4.PNG)
Am I doing something wrong?

**Comment by myleott** [comment]
Does it work with single GPU training? Try `CUDA_VISIBLE_DEVICES=0 python train.py (...)`

Also can you try doing `pip install fairseq` instead and see if it works with `CUDA_VISIBLE_DEVICES=0 fairseq-train (...)`?

**Comment by ajesujoba** [comment]
@myleott that works tho. I tried `CUDA_VISIBLE_DEVICES=0,2 python train.py (...)` and it also worked. But in the case of using all the GPUs on the node, it fails.

**Comment by nicola-decao** [comment]
@ajesujoba How did you solve? I am having the same problem.

**Comment by ajesujoba** [comment]
Hi @nicola-decao , using `CUDA_VISIBLE_DEVICES=0 fairseq-train (...)?` as suggested by @myleott worked for me.

**Comment by nicola-decao** [comment]
> Hi @nicola-decao , using `CUDA_VISIBLE_DEVICES=0 fairseq-train (...)?` as suggested by @myleott worked for me. Same applies for multi GPUs

Ho did this applied to multi GPUs? Didn't `CUDA_VISIBLE_DEVICES=0` triggers the use of only one GPU?

**Comment by ajesujoba** [comment]
@nicola-decao , for `CUDA_VISIBLE_DEVICES=0 fairseq-train (...)` it implies you are using the GPU with rank `0`, while `CUDA_VISIBLE_DEVICES=1 fairseq-train (...)` implies you are using GPU with rank 1. 

In the multi GPU case, `CUDA_VISIBLE_DEVICES=0,1,2 python train.py (...)` means you are using multi GPU setting with GPU ranks 0,1 and 2.

**Comment by nicola-decao** [comment]
> @nicola-decao , for `CUDA_VISIBLE_DEVICES=0 fairseq-train (...)` it implies you are using the GPU with rank `0`, while `CUDA_VISIBLE_DEVICES=1 fairseq-train (...)` implies you are using GPU with rank 1.
> 
> In the multi GPU case, `CUDA_VISIBLE_DEVICES=0,1,2 python train.py (...)` means you are using multi GPU setting with GPU ranks 0,1 and 2.

Yes I understand that, but I have multiple GPUs on multiple nodes. Should I use `CUDA_VISIBLE_DEVICES=0,..,N train.py (..)` where `N` is the number of GPUs on my nodes?

**Comment by ixlan** [comment]
I had a similar problem when I used a custom dataset that was converting raw data to tensors during reading. PyTorch as issues with tensors created on different processes, see links below.

https://discuss.pytorch.org/t/multiprocessing-using-torch-multiprocessing/11029
https://discuss.pytorch.org/t/using-torch-tensor-over-multiprocessing-queue-process-fails/2847/2

My solution was to convert data to tensors in the __getitem__ method instead.
